engine/pkg/dm/message_agent.go

```diff
@@ -0,0 +1,404 @@
+}
+
+// SendRequest send request synchronously.
+// caller should add its own retry mechanism if needed.
+// caller should persistent the request itself if needed.
+func (agent *MessageAgentImpl) SendRequest(ctx context.Context, senderID string, command string, req interface{}) (interface{}, error) {
+	sender, err := agent.getSender(senderID)
+	if err != nil {
+		return nil, err
+	}
+	ctx2, cancel := context.WithTimeout(ctx, defaultRequestTimeOut)
+	defer cancel()
+	log.L().Debug("send request", zap.String("sender-id", senderID), zap.String("command", command), zap.Any("req", req))
+	return agent.messagePair.sendRequest(ctx2, generateTopic(agent.id, senderID), command, req, sender)
+}
+
+// sendResponse send response asynchronously.
+func (agent *MessageAgentImpl) sendResponse(ctx context.Context, senderID string, msgID messageID, command string, resp interface{}) error {
+	sender, err := agent.getSender(senderID)
+	if err != nil {
+		return err
+	}
+	ctx2, cancel := context.WithTimeout(ctx, defaultResponseTimeOut)
+	defer cancel()
+	log.L().Debug("send response", zap.String("sender-id", senderID), zap.String("command", command), zap.Any("resp", resp))
+	return agent.messagePair.sendResponse(ctx2, generateTopic(agent.id, senderID), msgID, command, resp, sender)
+}
+
+// onMessage receive message/request/response.
+// Forward the response to the corresponding request request.
+// According to the command, the corresponding message processing function of defaultHandler will be called.
+// According to the command, the corresponding request processing function of defaultHandler will be called, and send the response to caller.
+// NOTE: comand name should same as handler name.
+// MessageFuncType: func(ctx context.Context, msg *interface{}) error {}
+// RequestFuncType(1): func(ctx context.Context, req *interface{}) (resp *interface{}, err error) {}
+// RequestFuncType(2): func(ctx context.Context, req *interface{}) (resp *interface{}) {}
+func (agent *MessageAgentImpl) onMessage(topic string, msg interface{}) error {
+	log.L().Debug("on message", zap.String("topic", topic), zap.Any("msg", msg))
+	m, ok := msg.(*message)
+	if !ok {
+		return errors.Errorf("unknown message type of topic %s", topic)
+	}
+
+	switch m.Type {
+	case responseTp:
+		return agent.handleResponse(m.ID, m.Command, m.Payload)
+	case requestTp:
+		sendID, _ := extractTopic(topic)
+		return agent.handleRequest(sendID, m.ID, m.Command, m.Payload)
+	default:
+		return agent.handleMessage(m.Command, m.Payload)
+	}
+}
+
+// handleResponse receive response.
+func (agent *MessageAgentImpl) handleResponse(id messageID, command string, resp interface{}) error {
+	handler := reflect.ValueOf(agent.commandHandler).MethodByName(command)
+	if !handler.IsValid() {
+		return errors.Errorf("response handler for command %s not found", command)
+	}
+	handlerType := handler.Type()
+	if handlerType.NumOut() != 1 && handlerType.NumOut() != 2 {
```

Comment:
there's a requirement for commandHandler, we should add the comment above NewMessageAgentImpl